### ISSUE TITLE
About new object evaluation

### ISSUE THREAD

**Issue by JasZhanAva** [GuessWhatGame/guesswhat#10]
Hi,

As I go through the code, the evaluation on test sets (new image) is perfect, however, I can not find the evaluation corresponding to the new object, which is referred as 'uniformly sampling objects
within the training set', I tried to directly replace the test set as train set, but I don't think that's it. Can you give some explanations about this? Thanks!

**Comment by fstrub95** [comment]
Sure!
In train_qgen_reinforce, you have the following batchifier/iterator/evaluator:

>             train_batchifier = LooperBatchifier(tokenizer, loop_sources, train=True)
>  
>             train_score = looper_evaluator.process(sess, train_iterator,
>                                                    optimizer=optimizer,
>                                                    mode="sampling")

This objects are going to compute the accuracy on new objects.
In a few words, the LooperBatchifier take the train_dataset as input.
It will then take create a new dataset by sampling all the images in the training dataset once (cf filter method)

At each turn the LooperBatchifier is going to randomly pick an object in the list of available objects (l52). 
random_index = random.randint(0, len(game.objects) - 1)

We finally report the accuracy over this newly generated games. (Generated on image on the training dataset but with new objects)

Thus, you can report the training accuracy of the network with the highest val accuracy or you can  put an aditionnal evaluation step at the end of the file.  

As the looper flag : "train=True/False" is a bit misleading. I replaced it with the flag "generate_new_games"
https://github.com/GuessWhatGame/guesswhat/commit/5c29897d0467b43359777065028d3a35e22a6873

Tell me if something is not clear.

KInd regards, 
Florian

**Comment by JasZhanAva** [comment]
Thanks, it worked.

**Comment by fstrub95** [comment]
I pushed a new version of train_reinforce.py that computes new_object by default. You also have a flag to use beam_search/greedy/sampling.
In other words, you can now genera the the full accuracy table on the fly.
It also dumps a file that you can use to compute all the figures in the CVPR paper

**Comment by JasZhanAva** [comment]
Thanks, I noticed the changes, will test them soon.

**Comment by JaceZhanAva** [comment]
I try to dump the dialogue with the flag=True, however, the dumped file is empty.

**Comment by JasZhanAva** [comment]
I try to dump the dialogue with the flag=True, however, the dumped file is empty.

**Comment by fstrub95** [comment]
I just tested it and it worked well...
Try to not put the flag to True as it is the default behavior

**Comment by fstrub95** [comment]
Do you still observe the issue? I cannot reproduce the bug on my side.

**Comment by JasZhanAva** [comment]
Sorry for the late reply, I debug the code, and it seems happens at the basic_looper file, line 84, it doesn't execute the loop. The store_game flag is True. I don't know whether it's my own problem, I'll look into it later. Thanks! :)

**Comment by fstrub95** [comment]
Did you pull the last version of the code?
https://github.com/GuessWhatGame/guesswhat/commit/9c9a477b8ce657eafb4291016fba2ddd9b4d0c98

**Comment by JasZhanAva** [comment]
I just checked, looks like my version is not the latest, there is no line 83 in my file. I'll update my code, thanks! :)

**Comment by JasZhanAva** [comment]
I'll close the issue for now. Thanks! :)
